### PR TITLE
[WIP] Add Tortoise TTS

### LINKED ISSUE #3891: Add Tortoise TTS as a pipeline
### Model/Pipeline/Scheduler description

TorToise is a multi-voice text-to-speech system, which describes a way to apply recent advances in the image generative domain to speech synthesis. It would be great to have this model in diffusers.

### Open source status

- [X] The model implementation is available
- [X] The model weights are available (Only relevant if addition is not a scheduler).

### Provide useful links for the implementation

Paper - https://arxiv.org/pdf/2305.07243.pdf
Github repo - https://github.com/neonbjb/tortoise-tts

@sanchit-gandhi @Vaibhavs10

### PR DESCRIPTION
**Pull request by susnato**
# What does this PR do?

Adds Tortoise TTS Pipeline and Fixes #3891 

Before Adding this pipeline, we need to make sure these two PR's are merged - 
- [ ] https://github.com/huggingface/transformers/pull/24745
- [ ] https://github.com/huggingface/transformers/pull/24799

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by susnato** [comment]
cc : @sanchit-gandhi @dg845

**Comment by dg845** [comment]
Hi, I pulled the [`tortoise_tts`](https://github.com/susnato/diffusers/tree/tortoise_tts) branch from `susnato/diffusers` and made my own local branch at [`dg845/diffusers`](https://github.com/dg845/diffusers/tree/tortoise_tts) with some changes, but I'm not sure how to make those changes show up on this PR.

(When I click the "Open pull request" button on GitHub, I'm not able to specify `susnato/diffusers` as the base repository, and I'm not sure setting `huggingface/diffusers` as the base repository will do the right thing [`tortoise_tts` isn't a branch on `huggingface/diffusers`].)

**Comment by susnato** [comment]
Hi @dg845, I just invited you to collaborate on my `susnato/diffusers`(you will get an email notification).  
Please accept that and clone the `susnato/diffusers` and then checkout the `tortoise_tts` branch, now after you make progress on this branch(on susnato/diffusers not on dg845/diffusers), please push it to my repository (it will be something similar to this - `git push susnato/diffusers tortoise_tts`). 

Since I gave you access to my repo, we can now both push changes to `tortoise_tts` of `susnato/diffusers` and the changes will show on this PR! 

Please let me know if this is working or not!

**Comment by susnato** [comment]
I have tested and added the resnet block, tomorrow will do the Attention block @dg845. Let's first add all the modules(those that you and I specified in slack) and create a basic diagram until the CLVP and Univnet is merged after that we can transfer the weights and finalize the whole model! 
Later we can focus on writing docs/tests.

**Comment by dg845** [comment]
> Please let me know if this is working or not!

Looks like it is working, was able to push a commit :).

**Comment by dg845** [comment]
I have moved the resnet block code to `modeling_tortoise_tts.py` and added some initial pipeline code in `pipeline_tortoise_tts.py`. I think keeping only pipeline code in `pipeline_tortoise_tts.py` and putting new module code in `modeling_*.py` files (like e.g. the [Versatile Diffusion pipeline](https://github.com/huggingface/diffusers/tree/main/src/diffusers/pipelines/versatile_diffusion)) will make the code more clear.

**Comment by susnato** [comment]
Just started the part of checkpoint conversion script for the diffusion decoder model, will also add the clvp model conversion script later(btw the weights loading code is unfinished and a pure mess, will update it in the next commit), also it seems that you have done a lot of work here! I need to catch up.

**Comment by dg845** [comment]
Just a heads up that I have refactored `modeling_tortoise_tts.py` into 3 files: 

- `modeling_common.py`: modules not used in autoregressive modeling or diffusion modeling, like `ConditioningEncoder` and `RandomLatentConverter`, as well as blocks shared in common like `AttentionBlock`
- `modeling_autoregressive.py`: modules used in autoregressive modeling, such as Tortoise TTS's version of GPT2 and blocks specific to autoregressive modeling (like the autoregressive version of [`ResBlock`](https://github.com/152334H/tortoise-tts-fast/blob/main/tortoise/models/autoregressive.py#L22))
- `modeling_diffusion.py`: modules used in diffusion modeling, such as the diffusion denoising model, and blocks specific to diffusion modeling (like the diffusion `ResnetBlock1D` block)

The code can probably be simplified further but I think this makes sense for now.

**Comment by susnato** [comment]
The diffusion decoder attention outputs are same now, the whole decoder model will probably be ready in next 1 or 2 days since the Resnet outputs are already verified to the official repo. Also should we not place `TortoiseTTSDiffusionModelAttention` in `modeling_diffusion.py` instead of `modeling_common.py`? if I am not wrong that module is specific to the decoder model only.

Please let me know what you think.

**Comment by dg845** [comment]
`ConditioningEncoder` also uses the same attention block as the diffusion denoising model (parallel to how both [`tortoise.models.autoregressive.ConditioningEncoder`](https://github.com/152334H/tortoise-tts-fast/blob/main/tortoise/models/autoregressive.py#L171) and [`tortoise.models.diffusion_decoder.DiffusionLayer`](https://github.com/152334H/tortoise-tts-fast/blob/main/tortoise/models/diffusion_decoder.py#L129) both use [`tortoise.models.arch_util.AttentionBlock`](https://github.com/152334H/tortoise-tts-fast/blob/main/tortoise/models/arch_util.py#L86) in the original code), so I think `TortoiseTTSDiffusionModelAttention` should go in `modeling_common.py`.

I guess one difference is that when `AttentionBlock` is used in autoregressive modeling the relative position embeddings aren't used, whereas it is always used in diffusion modeling. So we could potentially replace the attention block used in autoregressive modeling with a more "vanilla" attention block (possibly one already implemented in `diffusers`).

[For context, the only place `AttentionBlock` is used in autoregressive modeling is in the `ConditioningEncoder`. In the current design the `ConditioningEncoder` module is meant to work for both autoregressive and diffusion modeling; for the latter, it needs access to relative position embeddings, which is why it's intended to use `TortoiseTTSDiffusionModelAttention` currently.]

**Comment by susnato** [comment]
.

**Comment by manmay-nakhashi** [comment]
@susnato you can skip clvp if you want 
https://huggingface.co/spaces/Manmay/tortoise-tts refer the code in api.py over here

**Comment by patrickvonplaten** [comment]
Hey @susnato, 

Do you need help here? Would you like us to take over the PR?

**Comment by susnato** [comment]
Hi @patrickvonplaten ,  me and @dg845 are currently waiting for the [CLVP](https://github.com/huggingface/transformers/pull/24745) and [Univnet](https://github.com/huggingface/transformers/pull/24799) PRs to get merged. They are already approved by @ylacombe but waiting for a core-maintainer in the `transformers` repository to merge them.

A lot of work has already been done for this PR and we will need to incorporate some more changes before we finalize. If you want, we can get it ready in one week.

We would greatly appreciate if you could help us in getting `clvp` and `univnet` PRs merged as soon as possible.

**Comment by susnato** [comment]
Hi @manmay-nakhashi , thanks for the tip! 
BTW here we are adding the `autoregressive_model` of tortoise-tts alongside clvp in the clvp PR, so we need to merge it eventually.

**Comment by manmay-nakhashi** [comment]
@susnato yes , it's a good to have in pipeline but without that also it works.

**Comment by ylacombe** [comment]
Hey @susnato , I see that you are really active in integrating Tortoise!
Let me know if I can be of any help here, or if you need a first review!

**Comment by susnato** [comment]
Hi @ylacombe, sure I till let you know once it's finished, and sorry it's taking soo long.

**Comment by ylacombe** [comment]
Hey @tuanh123789, let's first ping @susnato to make sure he doesn't want or have bandwidth to finish this PR! 
If that's the case, feel free to continue the PR and ping me once you need a review!

@susnato, let us know ! You've already make a great effort on this PR, would you like to finish it up? Thanks!

**Comment by tuanh123789** [comment]
It seems like @susnato  is busy and no longer working on this PR. Perhaps I will continue and inherit his contributions to complete this pipeline @ylacombe

**Comment by susnato** [comment]
Hello, I am so sorry to everyone that I couldn't finish this PR 😢 . 

@tuanh123789 Please feel free to take this up, and as far as I remember me and @dg845 have already impelemented CLVP and UnivNet vocoder and I was able to get the logits from start to vocoder within 2e-2 atol (within acceptable range of diffusers as far as I am aware), so it needs few more work to make it e2e compatible.

I can also invite you to my diffusers branch so that you can continue the work from there (if you want of course)

Also maybe @ylacombe you could invite @tuanh123789 to our shared slack channel (if possible) so that he could get more idea of the current state and issues that we were facing.

**Comment by susnato** [comment]
@tuanh123789 Let me know if you need any more pointers, I will try to answer as much as I can!

**Comment by ylacombe** [comment]
Thanks for the update @susnato! @tuanh123789 feel free to reach out on X or LI to get you on the channel!

**Comment by tuanh123789** [comment]
> Hello, I am so sorry to everyone that I couldn't finish this PR 😢 .
> 
> @tuanh123789 Please feel free to take this up, and as far as I remember me and @dg845 have already impelemented CLVP and UnivNet vocoder and I was able to get the logits from start to vocoder within 2e-2 atol (within acceptable range of diffusers as far as I am aware), so it needs few more work to make it e2e compatible.
> 
> I can also invite you to my diffusers branch so that you can continue the work from there (if you want of course)
> 
> Also maybe @ylacombe you could invite @tuanh123789 to our shared slack channel (if possible) so that he could get more idea of the current state and issues that we were facing.

Sure, pls add me to your branch
